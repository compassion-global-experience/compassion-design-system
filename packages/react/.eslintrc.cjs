module.exports = {
  extends: [
    'standard-jsx',
    'standard-react',
    'plugin:jsx-a11y/recommended',
    'plugin:react/recommended',
    'plugin:react/jsx-runtime',
    'plugin:react-hooks/recommended',
    'plugin:storybook/recommended',
  ],
  env: {
    browser: true,
    es6: true,
    jest: true,
    node: false,
  },
  rules: {
    'jsx-quotes': ['error', 'prefer-double'],
  },
};
