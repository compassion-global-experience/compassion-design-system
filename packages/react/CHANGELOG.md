# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2023-04-08)

### Features

- **react:** css usage and documentation ([11e317c](https://github.com/compassion-global-experience/compassion-design-system/commit/11e317cca4ecdaa385e86d3a5571d7dca554685d))
- **react:** re-export css package ([67549a5](https://github.com/compassion-global-experience/compassion-design-system/commit/67549a58de3de6e02b58c5bdb6f4605001aae783))
- **react:** switch between themes in storybook ([79119ba](https://github.com/compassion-global-experience/compassion-design-system/commit/79119ba48ed3410eaebe9a91ccd076b05c2580c1))
