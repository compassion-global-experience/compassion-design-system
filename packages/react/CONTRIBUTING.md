# Contributing to the `react-web` Package

The `react` package is structured with a component-per-directory convention
inside the `src` directory. Each component directory generally contains 3 items:

- A **TSX file** that describes the component‚Äôs structure and behavior
- A **.stories.ts file** that illustrates most of the ways the component can
  be configured
- An **index.ts** file that exports the TSX component‚Äôs contents

## Workflow

The package uses [Storybook](https://storybook.js.org/docs/react/get-started/introduction)
to facilitate component development and documentation. Storybook uses the files
mentioned above to render components in their various states, and serves as
documentation of those components by surfacing code comments, `propTypes`, and
`defaultProps` values.

To launch Storybook, make sure you‚Äôre in the [/packages/react](/packages/react)
directory and run the command `npm run storybook`. After the project is built,
Storybook will launch in your default browser. New components or modifications
will be displayed in real time! üéâ

The configurations outlined in each component‚Äôs .stories.ts file can be seen in
Storybook‚Äôs **Docs** tab.

## Commit Messages & Versioning

Every package in the compassion-gds monorepo adheres to
[semantic versioning](https://semver.org/). In order to simplify the versioning
and release process, please be sure your commit messages adhere to the
[‚ÄúConventional Commits‚Äù pattern](https://www.conventionalcommits.org/en/v1.0.0/#summary).
You may wish to install the [commitizen/cz-cli](https://github.com/commitizen/cz-cli)
to make this near-bulletproof.

## Making a Pull Request

Once your changes are ready, open a pull request against the `development`
branch of the project, and it will be merged in as quickly as possible!
