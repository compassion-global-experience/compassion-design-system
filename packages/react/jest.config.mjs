/*
 * For a detailed explanation regarding each configuration property, visit:
 * https://jestjs.io/docs/configuration
 */

export default {
  // A preset that is used as a base for Jest's configuration
  preset: 'ts-jest/presets/default-esm',

  // Automatically clear mock calls, instances, contexts and results before every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  collectCoverage: true,

  // The directory where Jest should output its coverage files
  coverageDirectory: "coverage",

  // Indicates which provider should be used to instrument code for coverage
  coverageProvider: "v8",

  // The test environment that will be used for testing
  testEnvironment: "jsdom",

  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
  moduleNameMapper: {
    ".(css|less|scss)$": "identity-obj-proxy",
  },

  // Whether to use watchman for file crawling
  watchman: true,
};
