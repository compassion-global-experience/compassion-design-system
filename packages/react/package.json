{
  "name": "@compassion-gds/react",
  "version": "0.2.0-alpha.2",
  "description": "",
  "type": "module",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "files": [
    "dist"
  ],
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "browser": {
        "default": "./dist/esm/index.js"
      },
      "import": "./dist/esm/index.js",
      "require": "./dist/cjs/index.js"
    },
    "./css/base.css": "./dist/css/reset.css",
    "./css/light.css": "./dist/css/cds-light.css",
    "./css/dark.css": "./dist/css/cds-dark.css",
    "./css/*": "./dist/css/*"
  },
  "scripts": {
    "dev": "rollup -c -w",
    "copy-css": "bash ./scripts/copy-css.sh",
    "prepare": "rollup -c && npm run copy-css",
    "test": "jest",
    "storybook": "storybook dev -p 6007",
    "build:storybook": "storybook build --quiet",
    "percy": "cross-env PERCY_TOKEN=$PERCY_TOKEN_REACT percy storybook ./storybook-static",
    "percy-snapshots": "npm run build:storybook && npm run percy"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/compassion-global-experience/compassion-design-system.git",
    "directory": "packages/react"
  },
  "keywords": [
    "compassion",
    "compassion-gds",
    "gds",
    "react"
  ],
  "author": "Compassion XD",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/compassion-global-experience/compassion-design-system/issues"
  },
  "homepage": "https://github.com/compassion-global-experience/compassion-design-system#readme",
  "devDependencies": {
    "@babel/core": "^7.20.12",
    "@babel/preset-env": "^7.20.2",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "@compassion-gds/core": "^0.1.0-alpha.0",
    "@rollup/plugin-commonjs": "^24.0.1",
    "@rollup/plugin-node-resolve": "^15.0.1",
    "@rollup/plugin-replace": "^5.0.2",
    "@rollup/plugin-typescript": "^11.1.0",
    "@storybook/addon-actions": "^7.0.2",
    "@storybook/addon-docs": "^7.0.2",
    "@storybook/addon-essentials": "^7.0.2",
    "@storybook/addon-interactions": "^7.0.2",
    "@storybook/addon-links": "^7.0.2",
    "@storybook/preset-react-webpack": "7.0.2",
    "@storybook/react": "7.0.2",
    "@storybook/react-webpack5": "7.0.2",
    "@storybook/testing-library": "^0.0.14-next.1",
    "@testing-library/react": "^13.4.0",
    "@types/jest": "^29.4.0",
    "@types/react": "^18.0.27",
    "babel-loader": "^8.3.0",
    "eslint": "^8.33.0",
    "eslint-config-standard-jsx": "^11.0.0",
    "eslint-config-standard-react": "^13.0.0",
    "eslint-plugin-jsx-a11y": "^6.7.1",
    "eslint-plugin-react": "^7.32.2",
    "eslint-plugin-react-hooks": "^4.6.0",
    "eslint-plugin-storybook": "^0.6.10",
    "jest": "^29.4.1",
    "jest-environment-jsdom": "^29.4.1",
    "phosphor-react": "^1.4.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "rollup": "^3.20.2",
    "rollup-plugin-dts": "^5.1.1",
    "rollup-plugin-peer-deps-external": "^2.2.4",
    "rollup-plugin-postcss": "^4.0.2",
    "storybook": "7.0.2",
    "ts-jest": "^29.0.5",
    "webpack": "^5.75.0"
  },
  "peerDependencies": {
    "react": "16.8.x || 17.x || 18.x",
    "react-dom": "16.8.x || 17.x || 18.x"
  }
}
