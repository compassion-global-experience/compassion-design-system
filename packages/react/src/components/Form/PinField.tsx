'use client';

import {
  ChangeEvent,
  KeyboardEvent,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';
import styles from '@compassion-design-system/core/src/components/Form/pin.module.css';
import common from '@compassion-design-system/core/src/components/Form/form-common.module.css';

import { FieldSize, FieldState } from './Helpers';
import { getClasses } from '../../utils/classes';

type Values = string[];

interface PinProps {
  pattern?: RegExp;
  autoFocus?: boolean;
  secret?: boolean;
  state?: FieldState;
  onComplete?: (value) => void;
}

interface PinHookProps extends PinProps {
  values?: Values;
  onChange?: (values: Values) => void;
}

export interface PinFieldProps extends PinProps {
  label?: string;
  value?: string;
  length?: number;
  size?: FieldSize;
  className?: string;
}

const PinField = ({
  label,
  value,
  length = 4,
  state,
  size,
  secret = false,
  autoFocus = false,
  pattern = DEFAULT_PATTERN,
  className,
  onComplete,
}: PinFieldProps) => {
  const containerClassNames = getClasses(
    styles,
    ['cds-form__pin-field', `cds-form--${size}`],
    className,
  );
  const fieldClassNames = getClasses(styles, [
    'cds-form__field',
    `cds-form--${state}`,
  ]);
  const labelClassNames = getClasses(common, [
    'cds-form__label',
    `cds-form--${state}`,
  ]);

  const [values, setValues] = useState(() =>
    Array.from({ length }).map((_, i) => value?.toString()[i] || ''),
  );

  const fields = usePinHook({
    values,
    autoFocus,
    secret,
    pattern,
    state,
    onChange: (v) => setValues(v),
    onComplete,
  });

  return (
    <div className={getClasses(common, 'cds-form__field-row')}>
      {label && <label className={labelClassNames}>{label}</label>}
      <div className={containerClassNames}>
        {fields.map((fieldProps, index) => (
          <div key={index} className={fieldClassNames}>
            <input
              className={getClasses(styles, 'cds-form__input')}
              {...fieldProps}
            />
          </div>
        ))}
      </div>
    </div>
  );
};

PinField.displayName = 'PinField';

const usePinHook = ({
  values,
  autoFocus,
  secret,
  state,
  pattern,
  onChange: onChangeProp,
  onComplete,
}: PinHookProps = {}) => {
  const disabled = state === 'disabled';

  const fieldRefs = useRef<HTMLInputElement[]>(Array(values.length).fill(null));

  const setFocus = useCallback((index = 0) => {
    fieldRefs.current[index]?.focus();
  }, []);

  useEffect(() => {
    if (autoFocus) setFocus();
  }, [autoFocus, setFocus]);

  const setFieldRef = useCallback(
    (index: number) => (ref: HTMLInputElement) => {
      fieldRefs.current[index] = ref;
    },
    [],
  );

  const onChange = useCallback(
    (index: number) => (event: ChangeEvent<HTMLInputElement>) => {
      let { value } = event.target;

      value = value.trim();

      if (!pattern.test(value)) return;

      const nextValues = values.slice();
      nextValues[index] = value;
      onChangeProp(nextValues);

      if (value) {
        if (!nextValues.includes('')) {
          onComplete?.(nextValues.join(''));
        }

        if (index !== values.length - 1) {
          setFocus(index + 1);
        }
      }
    },
    [pattern, values, onChangeProp, onComplete, setFocus],
  );

  const onKeyDown = useCallback(
    (index: number) => (event: KeyboardEvent<HTMLInputElement>) => {
      if (event.key === 'Backspace' && !values[index] && index) {
        setFocus(index - 1);
      }
    },
    [values, setFocus],
  );

  const fields = values.map((value: string, index: number) => ({
    ref: setFieldRef(index),
    value,
    disabled,
    autoComplete: 'off',
    maxLength: 1,
    type: secret ? 'password' : 'text',
    ...(!disabled && {
      onChange: onChange(index),
      onKeyDown: onKeyDown(index),
    }),
  }));

  return fields;
};

const DEFAULT_PATTERN = /.*/;

export default PinField;
