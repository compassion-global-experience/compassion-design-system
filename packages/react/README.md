# GDS React Components

The Compassion Design System includes a streamlined set of React components for
fast, accessible development of apps and static sites.

## Installing and usage

To install the GDS React package use

```shell
npm install @compassion-gds/react
```

Or `yarn add @compassion-gds/react` if you're using yarn

Import the following css files from the entry point of your application

#### index.js
```js
// Replace {theme} with any of the available themes (e.g. light, dark)
import '@compassion-gds/react/css/base.css';
import '@compassion-gds/react/css/vars/{theme}.css';
```

GDS React components would follow the imported theme.

Sample component usage:

```jsx
import React from 'react';
import { Button } from '@compassion-gds/react';

export const PrimaryButton = (props) => (
  <Button primary size="small" {...props} />
);
```

## Themes

Available options:
- light: `@compassion-gds/react/css/light.css`
- dark: `@compassion-gds/react/css/dark.css`

## Contributing to the `react` Package

The `react` package is structured with a component-per-directory convention
inside the `src` directory. Each component directory generally contains 3 items:

- A **TSX file** that describes the component’s structure and behavior
- A **.stories.ts file** that illustrates most of the ways the component can
  be configured
- An **index.ts** file that exports the TSX component’s contents
