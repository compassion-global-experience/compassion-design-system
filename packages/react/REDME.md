# GDS React Components

The Compassion Design System includes a streamlined set of React components for
fast, accessible development of apps and static sites.

## Installing and usage

To install the GDS React package use

```shell
npm install @compassion-gds/react
```

Or `yarn add @compassion-gds/react` if you're using yarn  
(`@compassion-gds/css` is automatically installed as peer dependency)

Import the following css files from the entry point of your application

#### index.js
```js
// Replace {theme} with any of the available themes (e.g. light, dark)
import '@compassion-gds/css/reset.css';
import '@compassion-gds/css/vars/{theme}.css';
```

GDS React components would follow the imported theme.

Sample component usage:

```jsx
import React from 'react';
import { Button } from '@compassion-gds/react';

export const PrimaryButton = (props) => (
  <Button primary size="small" {...props} />
);
```

## Themes

Available options:
- light: `@compassion-gds/css/vars/light.css`
- dark: `@compassion-gds/css/vars/dark.css`
