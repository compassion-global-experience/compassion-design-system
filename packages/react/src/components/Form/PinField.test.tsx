import { render } from '@testing-library/react';

import PinField from './PinField';

describe('PinField', () => {
  it('should render `Small` size', () => {
    const { container } = render(<PinField label="Small" size="small" />);
    expect(container).toMatchSnapshot();
  });

  it('should render 8 inputs', () => {
    const { container } = render(<PinField label="Longer" length={8} />);
    expect(container).toMatchSnapshot();
  });

  it('should render `Error` state', () => {
    const { container } = render(<PinField label="Error" state="error" />);
    expect(container).toMatchSnapshot();
  });

  it('should render `Disabled` state', () => {
    const { container } = render(
      <PinField label="Disabled" state="disabled" />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render `Secret` mode (hidden values)', () => {
    const { container } = render(
      <PinField label="Secret" value="1234" secret />,
    );
    expect(container).toMatchSnapshot();
  });
});
