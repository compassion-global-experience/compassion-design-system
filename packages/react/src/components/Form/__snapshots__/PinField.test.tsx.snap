// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinField should render \`Disabled\` state 1`] = `
<div>
  <div
    class="form-field-row"
  >
    <label
      class="form-label disabled"
    >
      Disabled
    </label>
    <div
      class="pin-field  "
    >
      <div
        class="form-field disabled"
      >
        <input
          autocomplete="off"
          class="form-input"
          disabled=""
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field disabled"
      >
        <input
          autocomplete="off"
          class="form-input"
          disabled=""
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field disabled"
      >
        <input
          autocomplete="off"
          class="form-input"
          disabled=""
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field disabled"
      >
        <input
          autocomplete="off"
          class="form-input"
          disabled=""
          maxlength="1"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PinField should render \`Error\` state 1`] = `
<div>
  <div
    class="form-field-row"
  >
    <label
      class="form-label error"
    >
      Error
    </label>
    <div
      class="pin-field  "
    >
      <div
        class="form-field error"
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field error"
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field error"
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field error"
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PinField should render \`Secret\` mode (hidden values) 1`] = `
<div>
  <div
    class="form-field-row"
  >
    <label
      class="form-label "
    >
      Secret
    </label>
    <div
      class="pin-field  "
    >
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="password"
          value="1"
        />
      </div>
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="password"
          value="2"
        />
      </div>
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="password"
          value="3"
        />
      </div>
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="password"
          value="4"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PinField should render \`Small\` size 1`] = `
<div>
  <div
    class="form-field-row"
  >
    <label
      class="form-label "
    >
      Small
    </label>
    <div
      class="pin-field small "
    >
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`PinField should render 8 inputs 1`] = `
<div>
  <div
    class="form-field-row"
  >
    <label
      class="form-label "
    >
      Longer
    </label>
    <div
      class="pin-field  "
    >
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-field "
      >
        <input
          autocomplete="off"
          class="form-input"
          maxlength="1"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
